File: VB/DocumentManagerSerialization/ViewModels/DocumentViewModel.vb
Imports System
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm
Imports DevExpress.Mvvm.POCO
Imports System.Collections.Generic
Imports DocumentManagerSerialization.Common

Namespace DocumentManagerSerialization.ViewModels
    <POCOViewModel> _
    Public Class DocumentViewModel
        Implements ISupportLogicalLayout(Of ViewModelState), ISupportParentViewModel, IDocumentContent

        Public ReadOnly Property DocumentManagerService() As IDocumentManagerService Implements ISupportLogicalLayout(Of DocumentManagerSerialization.Common.ViewModelState).DocumentManagerService
            Get
                Return Me.GetService(Of IDocumentManagerService)()
            End Get
        End Property

        Public Overridable Property Caption() As String
        Public Overridable Property CanBeClosed() As Boolean
        Public Overridable Property State() As ViewModelState


        Public Sub New()
            State = New ViewModelState() With {.State = "Default State"}
        End Sub

        Public Sub OpenChildDocument()
            Dim document = DocumentManagerService.CreateDocument("DocumentView", Nothing, Me)
            document.DestroyOnClose = False
            document.Title = "Child Document"
            document.Id = "Child" & Guid.NewGuid().ToString().Replace("-", "")
            document.Show()
        End Sub


#Region "ISupportLogicalLayout"
        Public Function SaveState() As ViewModelState Implements ISupportLogicalLayout(Of DocumentManagerSerialization.Common.ViewModelState).SaveState
            Return State
        End Function

        Public Sub RestoreState(ByVal state As ViewModelState) Implements ISupportLogicalLayout(Of DocumentManagerSerialization.Common.ViewModelState).RestoreState
            Me.State = state
        End Sub


        Private ReadOnly Property ISupportLogicalLayout_CanSerialize() As Boolean Implements ISupportLogicalLayout.CanSerialize
            Get
                Return Not String.IsNullOrEmpty(State.State)
            End Get
        End Property

        Public ReadOnly Property LookupViewModels() As IEnumerable(Of Object) Implements ISupportLogicalLayout.LookupViewModels
            Get
                Return Nothing
            End Get
        End Property

#End Region

        #Region "ISupportParentViewModel"

        Public Overridable Property ParentViewModel() As Object Implements ISupportParentViewModel.ParentViewModel

        #End Region

        #Region "IDocumentContent"

        Public Property DocumentOwner() As IDocumentOwner Implements IDocumentContent.DocumentOwner

        Public Sub OnClose(ByVal e As System.ComponentModel.CancelEventArgs) Implements IDocumentContent.OnClose
            e.Cancel = Not CanBeClosed
        End Sub

        Public Sub OnDestroy() Implements IDocumentContent.OnDestroy
        End Sub

        Public ReadOnly Property Title() As Object Implements IDocumentContent.Title
            Get
                Return Caption
            End Get
        End Property

        #End Region
    End Class
End Namespace

File: CS/DocumentManagerSerialization/ViewModels/DocumentViewModel.cs
using System;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm;
using DevExpress.Mvvm.POCO;
using System.Collections.Generic;
using DocumentManagerSerialization.Common;

namespace DocumentManagerSerialization.ViewModels {
    [POCOViewModel]
    public class DocumentViewModel : ISupportLogicalLayout<ViewModelState>, ISupportParentViewModel, IDocumentContent {

        public IDocumentManagerService DocumentManagerService { get { return this.GetService<IDocumentManagerService>(); } }
        
        public virtual string Caption { get; set; }
        public virtual bool CanBeClosed { get; set; }
        public virtual ViewModelState State { get; set; }


        public DocumentViewModel() {
            State = new ViewModelState() { State = "Default State" };
        }

        public void OpenChildDocument() {
            var document = DocumentManagerService.CreateDocument("DocumentView", null, this);
            document.DestroyOnClose = false;
            document.Title = "Child Document";
            document.Id = "Child" + Guid.NewGuid().ToString().Replace("-", "");
            document.Show();
        }

        public ViewModelState SaveState() {
            return State;
        }

        public void RestoreState(ViewModelState state) {
            State = state;
        }

        #region ISupportLogicalLayout

        bool ISupportLogicalLayout.CanSerialize {
            get { return !String.IsNullOrEmpty(State.State); }
        }

        public IEnumerable<object> LookupViewModels {
            get { return null; }
        }

        #endregion

        #region ISupportParentViewModel

        public virtual object ParentViewModel { get; set; }

        #endregion

        #region IDocumentContent

        public IDocumentOwner DocumentOwner { get; set; }

        public void OnClose(System.ComponentModel.CancelEventArgs e) {
            e.Cancel = !CanBeClosed;
        }

        public void OnDestroy() { }

        public object Title {
            get { return Caption; }
        } 

        #endregion
    }
}